pkg/ddl/000030_commits_parents_index.up.sql

```diff
@@ -0,0 +1,4 @@
+BEGIN;
+CREATE INDEX graveler_commits_idx_parents ON graveler_commits(parents);
```

Comment:
I re-read https://stackoverflow.com/q/4058731/192263 and I am still not convinced that this index can ever be used.

Reply:
I tried it out.  According to this I can disable sequential scans, _except_ when they are the only way to perform a query:

> When indexes are not used, it can be useful for testing to force their use. There are run-time parameters that can turn off various plan types (see Section 19.7.1). For instance, turning off sequential scans (enable_seqscan) and nested-loop joins (enable_nestloop), which are the most basic plans, will force the system to use a different plan. If the system still chooses a sequential scan or nested-loop join then there is probably a more fundamental reason why the index is not being used; for example, the query condition does not match the index. (What kind of query can use what kind of index is explained in the previous sections.)

(https://www.postgresql.org/docs/13/indexes-examine.html)

And it turns out that an index on an array value is _not_ used for matching individual array elements:

```sql
ariels=# CREATE TABLE family (name TEXT, parents TEXT[]);
CREATE TABLE
ariels=# CREATE INDEX idx_parents ON family(parents);
CREATE INDEX
ariels=# SET enable_seqscan TO off;
SET
ariels=# INSERT INTO family VALUES('a', '{a,b,c}');
INSERT 0 1
ariels=# INSERT INTO family VALUES('b', '{c,b,f}');      
INSERT 0 1
ariels=# INSERT INTO family VALUES('c', '{c,x,f}');  
INSERT 0 1

ariels=# SELECT * FROM family WHERE parents[2] = 'b';
 name | parents 
------+---------
 a    | {a,b,c}
 b    | {c,b,f}
(2 rows)

ariels=# EXPLAIN ANALYZE SELECT * FROM family WHERE parents[2] = 'b';  -- this will seq scan (no index)
                                                       QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Seq Scan on family  (cost=10000000000.00..10000000001.04 rows=1 width=64) (actual time=40.776..40.780 rows=2 loops=1)
   Filter: (parents[2] = 'b'::text)
   Rows Removed by Filter: 1
 Planning Time: 0.171 ms
 JIT:
   Functions: 2
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 0.792 ms, Inlining 15.325 ms, Optimization 17.722 ms, Emission 7.617 ms, Total 41.455 ms
 Execution Time: 41.629 ms
(9 rows)


ariels=# CREATE INDEX idx_parents_2 ON family((parents[2]));
CREATE INDEX
ariels=# EXPLAIN ANALYZE SELECT * FROM family WHERE parents[2] = 'b';  -- This will index scan
                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Index Scan using idx_parents_2 on family  (cost=0.13..8.15 rows=1 width=64) (actual time=0.020..0.023 rows=2 loops=1)
   Index Cond: (parents[2] = 'b'::text)
 Planning Time: 0.103 ms
 Execution Time: 0.049 ms
(4 rows)
```